# U5_hypothesis
This depository supports a study of pre-mRNA splicing, a routine molecular process in the cell nucleus.

The code in this depository supports a study of pre-mRNA splicing, a routine molecular process in the cell nucleus.

Substitutions of the conserved +5G and -3C in human introns are compensated by Watson-Crick pairs between the exons and U5 snRNA  

Olga Artemyeva-Isman, Andrew C.G. Porter

Gene Targeting Group, Centre for Haematology, Department of Immunology and Inflammation, Imperial College Faculty of Medicine, Du Cane Road, London W12 0NN, UK


Key words: splice sites, splicing mutations, U5 snRNA, U6 snRNA, U2 snRNA, U1 snRNA, Group II intron retrotransposition, RNA base pair geometry 

Running title: The U5 hypothesis

Correspondece: Olga Isman, o.isman@imperial.ac.uk


Contribution to the field of RNA biology

Almost every human protein-coding transcript undergoes alternative splicing, made possible by the imperfect conservation of splice sites. The challenge for the spliceosome is to combine flexibility with routine precision. The recognition of splice sites is a multistage process which involves base-pairing with small nuclear (sn)RNAs, the components of the spliceosome. The intron interactions with U6 and U2 snRNAs are better studied than the exon interactions with U5 snRNA. We propose a new model for U5 binding to exons by comparison of splicing with retrotransposition, as the human spliceosome is essentially a ribozyme that shares a primordial mechanism of splicing catalysis with bacterial mobile introns. Statistical analysis of U5 interactions with human exons lends support to this model. We find that U5 Watson-Crick pairs with the exons show a clear pattern of compensation for substitutions of the conserved nucleotides in human introns, indicating a collective mechanism whereby U5, U6 and U2 recognise their variable binding sites. We suggest that snRNAs in the pre-catalytic spliceosome combine to ensure fidelity before the committed ribozyme core is configured. The new U5 model explains the effect of human mutations on splicing. We discuss potential applications for snRNA therapeutics and gene repair by reverse splicing.
